align_system/algorithms/decision_flow_adm/utils.py

```diff
@@ -0,0 +1,183 @@
+"""Utility functions for DecisionFlow ADM components."""
+
+import json
+import re
+from typing import Any, Dict, Optional
+from align_system.utils import logging
+
+log = logging.getLogger(__name__)
+
+
+def run_structured_inference_with_retry(
+    inference_engine,
+    dialog_prompt: str,
+    output_schema: Any,
+    stage_name: str,
```

Comment:
In general I like to give each function the minimum amount of information possible for it to do it's job.  I don't think it's worth passing in the step name _only_ for logging purposes.  I'm assuming the other logging messages around it and the stacktrace (if it fails) would be enough to identify which step the log messages are for.  Or you could add another log message right before this is called, making it explicit that you're attempting (un)structured inference with retries for step "x".